Multi-stock proposal

IvanChepurnyi opened the pull request:
This is not an actual pull request for Magento 2.0. It has been created to review the changes, before doing the final proposal to Magento Core.

IvanChepurnyi committed:
Warehouses, Inventories and Inventory registries

IvanChepurnyi committed:
Add inventory validator, inventory response, reservation management
Changed some method names, to remove overuse of inventory word
Added indexers

IvanChepurnyi committed:
Add inventory and record providers

vkorotun commented on app/code/Magento/CatalogInventory/Api/Data/WarehouseRecordInterface.php:
```diff
@@ -0,0 +1,115 @@
+     *
+     * @return int
+     */
+    public function getInventoryId();
```
Please, clarify: is it really supposed to be Inventory Id here, or Warehouse Id?

IvanChepurnyi replied:
@vkorotun probably it is better to call it warehouse id, as warehouse record is a backend representation of inventory record, so developers won't get confused.

vkorotun commented on app/code/Magento/CatalogInventory/Api/Data/InventoryInterface.php:
```diff
@@ -0,0 +1,62 @@
+ *
+ * @api
+ */
+interface InventoryInterface extends LocationInformationInterface
```
Please, clarify: how InventoryInterface and WarehouseInterface are connected (or they are not supposed to be)?

IvanChepurnyi replied:
WarehouseInterface is a backend representation of inventory. Inventory is meant to be used only on frontend and warehouse is going to be used to manage inventory items. Their identifiers are connected. Advice to separate them by purpose was given by Anton Krill before I started creating interfaces, and I liked the idea of the immutable object on the frontend, so the contract is neat. In this case, third-party is not bound to implemented mutation contract if warehouse management logic is different from core implementation.

IvanChepurnyi replied:
So to summarize the logic:

| Entity Description | Interface for Configuring Data in Backend | Interface for Business Logic |
| --- | --- | --- |
| Scope of product inventory availability | **WarehouseInterface**, _WarehouseConfigurationInterface_* | **InventoryInterface**, **InventoryConfigurationInterface** |
| Availability of product in particular scope | **WarehouseRecordInterface** | **InventoryRecordInterface** / **InventoryIndexRecordInterface** |
| Product Inventory Reservation | No stock reservation is manageable manualy | **InventoryReservationManagementInterface** creates and manages all **InventoryReservationInterface** instances to process mutations of stock on frontend. Idea that you cannot directly modify **InventoryRecordInterface**, only via **InventoryReservationInterface**. |

IvanChepurnyi replied:
@vkorotun let me know what do you think about this.

maghamed commented on app/code/Magento/CatalogInventory/Api/Data/WarehouseRecordInterface.php:
```diff
@@ -0,0 +1,115 @@
+     *
+     * @return InventoryInterface
+     */
+    public function getInventory();
```
Seems this method should be renamed to getWarehouse() which returns an instance of WarehouseInterface as well (according to current logic)

For now we have Warehouse and Inventory set of interfaces very similar to each other. For example, Warehouse interface has all getters provided by Inventory interface. The only difference that Warehouse interfaces are mutable, but inventory - not. 
Why not to create a relation between these kind of interfaces in code? For example. Read-only interfaces of Inventory, and Warehouse interfaces extends (inherit from) them and add mutability logic?

Because for now, if we will decide to add some logic to InventoryInterface in future we will have to add the same methods for WarehouseInterface as well. And this is obscure, because there is no contract in the code.

IvanChepurnyi replied:
@maghamed:
I wanted to avoid extending warehouse directly from it, as the business logic of inventory contract must not be duplicated in warehouse entities, which are used for managing first ones. Maybe it makes sense to create something like the following:

InventoryInterface -> InventoryDataInterface 
WarehouseInterface -> InventoryDataInterface

In this case, InventoryDataInterface contains all getter methods, that are shared between warehouse entity and inventory entity.

@maghamed, @vkorotun 
Also after taking a look one more time at my proposal I want to optimize similar logic of configuration retrieval by introducing ConfigurationAwareInterface. It should contain  getConfiguration() contract logic for InventoryInterface, InventoryRecordInterface.
And on the admin part:
Both WarehouseConfiguration and WarehouseRecordConfiguration will be based
BaseConfigurationInterface, that contains same setters and getters for both contracts.

maghamed replied:
{quote}
Maybe it makes sense to create something like the following:

InventoryInterface -> InventoryDataInterface 
WarehouseInterface -> InventoryDataInterface

In this case, InventoryDataInterface contains all getter methods, that are shared between warehouse entity and inventory entity.
{quote}
Yes, this looks reasonable for me.

And seems we need to do the same for interfaces:
WarehouseRecordInterface
InventoryRecordInterface
because these interfaces looks identical as well

vkorotun commented on app/code/Magento/CatalogInventory/Api/InventoryRegistryInterface.php:
```diff
@@ -0,0 +1,73 @@
+     * @param int|null $storeId
+     * @return InventoryInterface
+     */
+    public function getInventory($storeId = null);
```
I believe we want to eliminate dependency on store id and introduce some "virtual" scope identifier which might be configured per instance. Have you considered such option?

IvanChepurnyi replied:
Yes, as Inventory now do not have store id binded to it, but hovewer it should be possible to specify default inventory for a store, that's why there is option store parameter specified for it. Take a look at this chart to get the idea:
![Usage of Intentory Registry](https://www.lucidchart.com/publicSegments/view/55e5dcd3-8174-4289-9c64-71df0a00519d/image.png)

vkorotun replied:
Then you'd better to pass \Magento\Framework\App\ScopeInterface here.

IvanChepurnyi replied:
@vkorotun done

IvanChepurnyi committed:
Removed redundant configuration contracts for business logic
Added shared contracts for Inventory and Warehouse entities, to follow DRY principle
Added more specific methods for warehouse entities
Fixed some stylings in error interface

IvanChepurnyi committed:
Changed $storeId in InventoryRegistryInterface to use \Magento\Framework\App\ScopeInterface as parameter instead of $storeId
Removed InventoryRecordProviderInterface and InventoryProviderInterface as redundant

IvanChepurnyi committed:
Fix PHPDoc blocks for method, to reflect more correct flow according to chart

IvanChepurnyi commented:
Closed as of https://github.com/magento/magento2/pull/1790
